Core/src/main/java/com/plotsquared/core/queue/BlockArrayCacheScopedQueueCoordinator.java

```diff
@@ -0,0 +1,169 @@
+ *      | |__) | | ___ | |_| (___   __ _ _   _  __ _ _ __ ___  __| |
+ *      |  ___/| |/ _ \| __|\___ \ / _` | | | |/ _` | '__/ _ \/ _` |
+ *      | |    | | (_) | |_ ____) | (_| | |_| | (_| | | |  __/ (_| |
+ *      |_|    |_|\___/ \__|_____/ \__, |\__,_|\__,_|_|  \___|\__,_|
+ *                                    | |
+ *                                    |_|
+ *            PlotSquared plot management system for Minecraft
+ *               Copyright (C) 2014 - 2022 IntellectualSites
+ *
+ *     This program is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.plotsquared.core.queue;
+
+import com.plotsquared.core.location.Location;
+import com.sk89q.jnbt.CompoundTag;
+import com.sk89q.worldedit.function.pattern.Pattern;
+import com.sk89q.worldedit.math.BlockVector3;
+import com.sk89q.worldedit.world.biome.BiomeType;
+import com.sk89q.worldedit.world.block.BaseBlock;
+import com.sk89q.worldedit.world.block.BlockState;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * {@link QueueCoordinator} that caches all blocks set to it in a given array of form BlockState[][][]. An offset can be
+ * applied to blocks set to it, and the scope limited.
+ */
+public class BlockArrayCacheScopedQueueCoordinator extends ScopedQueueCoordinator {
+
+    private final BlockState[][][] blockStates;
+    private final int height;
+    private final int width;
+    private final int length;
+    private final int minY;
+    private final int maxY;
+    private final int scopeMinX;
+    private final int scopeMinZ;
+    private final int scopeMaxX;
+    private final int scopeMaxZ;
+
+    private int offsetX = 0;
+    private int offsetZ = 0;
+
+    /**
+     * Construct a new instance
+     *
+     * @param blockStates Array of form BlockState[y][x][z]. Must be fully initialised.
+     * @param minY        Minimum applicable y value. Used to account for negative world heights (e.g. -64). Inclusive
+     * @param min         Inclusive location of the minimum point to limit the scope to.
+     * @param max         Exclusive location of the maximum point to limit the scope to.
+     * @since TODO
+     */
+    public BlockArrayCacheScopedQueueCoordinator(@NonNull BlockState[][][] blockStates, int minY, Location min, Location max) {
```

Comment:
Passing the array is a little bit odd, it allows outside modification of an internal detail. Shouldn't it be enough to have min and max locations to create the array in the constructor? (note that the javadocs would need to be changed accordingly if this is changed). The name of the class could then also be changed to something more generic.

Reply:
The class has a very specific use in the first place. It was done to allow the array be filled by default with air (which would not be ideal for a generic queue class) as per the analyzeRegion method in HybridUtils. I suppose the oldBlocks variable doesn't necessarily need to be available there, but performance is increased at effectively no memory cost by simply referencing the array directly without checking if the coordinates are within range every single time (we know they are because that's the coordinates we iterate over in analyzeRegion). Keeping this performance improvement would require a method to retrieve the array cache from the queue anyway, and unless we add the extra overhead of cloning the entire BlockState[][][] array, it is still mutable at that point anyway.

Given generation would actually set air to the entire array anyway (and a null check can always be added to assume air), it probably would be better to have a method to simply return the array (as in the ChunkQueueCoordinator class that was used initially).

Though thinking of it, there is no reason why we couldn't just analyze chunk-by-chunk as the chunks in the world are loaded, using a off-the-main-thread single-threaded executor...

Reply:
Scrap that, we need to full data at once in order to properly analyse the faces